[로또] 지토 미션 제출합니다.

stopsilver13 opened the pull request.

Commit:
Make a TODO list draft for Lv.1 at README

Commit:
Edit TODO list to redistribute responsibility

Commit:
[FEATURE] Implement money input view and Add custom exception for money

inputMoney 구현 : 유효한 값이 입력될 때까지 입력받는 로직
> Money 객체 생성
> InvalidPaymentException 생성

Commit:
[FEATURE] Catch exception at Money's constructor

Money 예외처리 구현 + 테스트코드 작성
> StringUtils 임포트
> 빈문자열/공백 입력시 예외처리
> 숫자가 아닌 값 입력시 예외처리
> 천원~십만원 아닌 값 입력시 예외처리

Commit:
[FEATURE] Implement calculate buyable lotto number method at Money

Money : 구매 가능한 로또 갯수 구하는 메서드 작성
+ 상수 변수 처리

Commit:
Edit TODO list to change getting lotto numbers logic

Commit:
[FEATURE] Implement randomly get six numbers

랜덤하게 로또 숫자들을 가져오는 로직 구현 + 테스트코드 작성
> LottoNumber, LottoNumbers 객체 생성
> LottoNumbers 생성시 1~45의 값을 가지는 LottoNumber 생성

Commit:
[FEATURE] Implement generate lotto automatically

입력받은 갯수만큼 자동 로또 생성 로직 구현 + 테스트코드 작성
> Lotto, Lottos 객체 생성
> LottoCreator 인터페이스 생성 - AutoLottoCreator 생성

Commit:
[REFACTOR] Rename Money method to avoid confusion

Money : getBuyableLottoNumber() -> getBuyableLottoQuantity()

Commit:
Edit TODO list to redistribute responsibility

Commit:
[FEATURE] Implement winning lotto input view and Add custom exception for lotto

inputWinningLotto 구현 : 유효한 값이 입력될 때까지 입력받는 로직
> Money 객체 생성
> InvalidLottoNumbersException 생성
> LottoNumbers에 커스텀 LottoNumber 객체들을 가져올 수 있는 메서드 작성

Commit:
[FEATURE] Catch exception at WinningLotto's constructor

WinningLotto, Lotto, LottoNumbers 예외처리 구현 + 테스트코드 작성
> 빈문자열/공백 입력시 예외처리
> 숫자가 아닌 값 입력시 예외처리
> 입력한 숫자 갯수가 6개가 아닐시 예외처리
> 입력한 숫자가 로또에 사용되는 숫자가 아닐시 예외처리

Commit:
[REFACTOR] Combine LottoNumber and LottoNumbers to reuse lotto number

Commit:
Reflect done work at TODO list

Commit:
[FEATURE] Implement match lotto's Rank

로또의 당첨 결과 RANK를 가져오는 로직 구현
> Rank enum 생성
> InvalidRankMatchException 생성

Commit:
Edit TODO list to redistribute responsibility

Commit:
[FEATURE] Implement lottos result and calculate ROI

전체 로또의 결과를 구하고 ROI를 계산하는 로직 구현
> LottosResult 생성

Commit:
[REFACTOR] Fix bug from Combine LottoNumber and LottoNumbers at LottoCreator

수동 로또 생성 로직 구현
> 이전에 주석처리 한 코드가 실수로 커밋 됨

Commit:
[REFACTOR] Fix bug that throwing any exception at ManualLottoCreator

Commit:
[FEATURE & REFACTOR] Implement LottosResult testcod and wrap in package

LottosResult 테스트코드 작성
> 총 상금 계산
> ROI 계산

Commit:
Make a TODO list draft for Lv.2, 3 at README

Commit:
[FEATURE] Implement bonus number input view at winning lotto input view

> inputWinningLotto에 보너스 번호를 받는 기능을 추가
> inputwinningLotto의 로또 번호와 보너스 번호 input 기능 분리
> WinningLotto에 bonusNumber 프로퍼티 추가 & 생성자 수정

Commit:
[FEATURE] Implement manual lotto number input view and Add custom exception

inputManualLotto 구현 : 유효한 값이 입력될 때까지 입력받는 로직
> InvalidInputException 생성
> 수동 로또 숫자들을 받아 ManualLottoCreator 생성하는 로직

Commit:
[FEATURE] Reflect bonus number property in WinningLotto

WinningLotto 수정 + 테스트 코드 수정
> 보너스 숫자가 이전에 입력된 로또 숫자와 중복되는지 검증하는 로직

Commit:
[FEATURE] Reflect bonus number at Rank enum and Rank match logic

Commit:
Reflect done work at TODO list

Commit:
[FEATURE] Implement LottoApp, Outputview

> LottoFactory 생성

Commit:
[REFACTOR] Supplement test code and categorize class

phs1116 commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,47 @@
+            return MISS;
+        }
+
+        for (Rank rank : Rank.values()) {
```
`객체지향 생활원칙 규칙 1: 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.`
를 지켜서 하나의 인덴트로 표현해보면 어떨까요?
stream을 쓸수도 있겠네요. :)

stopsilver13 replied:
아! 그러네요 ^_^;
말씀하신대로 stream을 써서 바꿔봤는데 제가 stream이 아직 익숙하지 않아서 그런지 코드가 그닥 맘에 들지는 않네요ㅠㅠ 가독성이 더 좋은지도 잘 모르겠구요ㅠㅠ

phs1116 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,69 @@
+        return numbers.get(number);
+    }
+
+    public static List<LottoNumber> getLottoNumbers() {
```
해당 부분에 대해서 테스트가 진행되지 않은 것 같아요.
반환되는 값에 대한 검증을 테스트 할 수 있는 방법이 없을까요?
어렵다면 왜 어려울지, 어떻게 하면 테스트하기 쉬울지도 고민해보면 좋을 것 같습니다.

제 생각에 대해 말씀드리자면 이런 자동 생성에 대한 역할은 별도 factory로 분리되고, 추상화되어 관리되면 좋을 것 같네요.

stopsilver13 replied:
테스트 부분도 최대한 할 수 있는 만큼 구현해보았습니다!
그런데 자동 로또 숫자 생성 로직은 테스트하는 방법에 있어 임의로 두 개를 생성하고 두개가 다른지 비교하는 방법외에 테스트할 수 있는 방법이 또 있을까요...?

phs1116 commented on src/main/java/lotto/domain/creator/ManualLottoCreator.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class ManualLottoCreator implements LottoCreator {
```
Factory라고 보기에는 그저 문자열을 parasing하는 역할로 보여요.
두개의 책임을 분리해볼수 있지 않을까요?

stopsilver13 replied:
음 먼저 이부분에 대해 여쭤보고 싶은 것이 있습니다. 제가 이렇게 한 이유는 자동으로 로또를 생성하는 경우 인자를 따로 받을 필요가 없지만 수동으로 로또를 생성하는 경우 어떤 번호로 로또를 만들 것인지에 대한 List<String> 인자를 받아야 합니다. 하지만 같은 인터페이스를 상속(?)하는 경우 인자를 서로 다르게 구현할 수 없서 생성자 부분에서 이를 인자로 받아서 로또 생성에 필요한 숫자들을 저장해주는 역할을 맡겼습니다.

일단은 지금 뭔가 LottoCreator랑 LottoFactory랑 역할이 잘 배분되지 않아 생기는 문제같기는 해서^_^; 더 고민해보겠습니다. 이부분은 구현하면서 여쭤보고 싶었던 부분인데 마침 이야기 해주셔서 글 남깁니다!

stopsilver13 replied:
parsing 하는 부분은 CustomStringUtils 메서드로 해결이 되어서 그렇게 했고, input 값 다듬는 부분은 Creator에서 Factory로 역할을 넘겼습니다! 하고 보니 쪼금은 알것 같긴 합니다!

phs1116 commented on src/main/java/lotto/domain/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+public class WinningLotto {
+    private final Lotto lotto;
```
컴포지션 👍

phs1116 commented on src/main/java/lotto/domain/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,71 @@
+        }
+    }
+
+    private List<Integer> toInt(List<String> strings) {
```
일반적으로 toInt와 같이 to ~형 메서드는 호출되는 객체 자체를 다른 형태로 바꾸는(conversion)하는 메서드 이름으로 사용됩니다.
다른 적당한 이름으로 수정해보면 좋을 것 같아요. :)

stopsilver13 replied:
그렇군요, parseInts로 변경하였습니다!

phs1116 commented on src/main/java/lotto/domain/lotto/WinningLotto.java:
```diff
@@ -0,0 +1,71 @@
+        return numbers;
+    }
+
+    private int parseInt(String input) {
```
파싱 하는 역할에 대해서는 별도 책임으로 분리 할 수 있을 것 같습니다.

stopsilver13 replied:
공백인지 체크하는 역할과 파싱하는 역할을 CustomStringUtils 객체로 묶어 분리하였는데요, 여기서 고민이 많이 생겼습니다. 제가 이번에 커스텀 Exception을 만들어서 구체적으로 예외처리를 해주면 좋다고 배웠는데요, 이러한 기능이 분리되다 보니 한 객체를 만들면서 발생할 수 있는 Exception이 여러개가 되더라고요ㅠㅠ 그러면 Inputview에서 예외를 받을 때 구체적으로 받을 수 없고 Exception 이렇게 두루뭉실하게 받아야 하는데 이것도 좋은 방법이 아니라고 피드백을 예전에 받은 적이 있습니다. 저는 Exception으로 범용적으로 예외를 받는 것에 대해 그리 거리낌이 없어 일단은 그렇게 구현했는데, 이러한 경우 리뷰어님이라면 어떻게 하실것 같으신가요?

phs1116 commented on src/main/java/lotto/domain/lotto/LottoFactory.java:
```diff
@@ -0,0 +1,11 @@
+import java.util.List;
+
+public class LottoFactory {
+    public static List<Lotto> createLottoList(int lottoQuantity, LottoCreator creator) {
```
LottoFactory의 역할이 무엇인지, 왜 있어야하는 것인지 명확하지 않은 것 같습니다. :)
LottoCreator는 숫자만 만들어주는 역할이고, LottoFactory는 로또를 만들어주는 역할이라면 납득이 갈 것 같은데
LottoCreator자체가 팩토리의 역할을 하고 있네요!

stopsilver13 replied:
맞습니다ㅠㅠ 사실 만들면서도 조금 이상하다 싶었는데 막막해서 일단은 리뷰요청 드렸습니다. 이부분 고민해서 Creator와 Factory가 적절히 역할을 나눠 갖도록 해보았습니다.

phs1116 reviewed (changes requested):
안녕하세요! 리뷰어 화투입니다.
객체간의 책임 분리 👍 
전체적으로 매우 잘 구현해주신 것 같아요. 👍 
피드백 몇가지 추가했는데요, 확인 부탁드릴게요!
궁금하신 점 있으시면 언제든지 DM주세요!

Commit:
[REVIEW] Reduce depth by using stream

> Rank enum 의 valueOf 메서드

Commit:
[REVIEW] Rename method

> WinningLotto 의 toInt -> parseInts

Commit:
[REVIEW] Separate method about string as a CustomStringUtils

> checkIsBlank, parseInts, parseInt
> Money, WinningLotto, ManualLottoCreator, InputView 에 반영

Commit:
[REVIEW] Redistribute responsibility to LottoCreator and LottoFactory

Commit:
[REFACTOR] Fix bug that throwing any exception when input is blank

> InputVeiw inputManualLottoNumber 메서드 공백 입력시 예외 처리

phs1116 commented on src/main/java/lotto/domain/LottosResult.java:
```diff
@@ -0,0 +1,50 @@
+public class LottosResult {
+    private static final int INITIAL_NUMBER = 0;
+
+    private final EnumMap<Rank, Integer> lottosResult = new EnumMap<>(Rank.class);
```
`이펙티브 자바 아이템64 객체는 인터페이스를 사용해 참조하라`
참고 부탁드릴게요 :)

stopsilver13 replied:
아하 그렇군요! 반영하였습니다.

phs1116 commented on src/main/java/lotto/domain/LottosResult.java:
```diff
@@ -0,0 +1,50 @@
+
+        for (Lotto lotto : myLottos.getLottos()) {
+            Rank rank = winningLotto.matchRank(lotto);
+            lottosResult.put(rank, lottosResult.get(rank) + 1);
```
map.getOrDefault()를 사용하면 setUpResult를 사용하지 않을수도 있을 것 같네요. :)

stopsilver13 replied:
그렇네요! 감사합니다!

phs1116 commented on src/main/java/lotto/domain/lotto/LottoFactory.java:
```diff
@@ -0,0 +1,38 @@
+        List<Lotto> lottos = new ArrayList<>();
+
+        for (String input : inputs) {
+            ManualLottoCreator creator = new ManualLottoCreator(CustomStringUtils.parseInts(input));
```
`이펙티브 자바 아이템64 객체는 인터페이스를 사용해 참조하라` 참고 부탁드립니다.
그리고 구현체에 직접 의존하고 있는데요,
개방 폐쇄 원칙(OCP), 의존성 역전 원칙(DIP)에 위반하고 있는 것 같아요!
인터페이스를 따로 뺀 의미가 많이 사라지는 것 같습니다. :)

외부에 의해서 주입을 받아서 사용하는 형태로 느슨하게 가져가는게 좋을 것 같아요.
그리고 이를 잘 활용해보시면 두개의 메서드로 나누지 않고도 제공할 수 있을것 같네요. :)

phs1116 commented on src/main/java/lotto/LottoApp.java:
```diff
@@ -0,0 +1,37 @@
+
+        OutputView.printPurchaseResult(manualLottoQuantity, autoLottoQuantity);
+
+        totalLottos.addAll(manualLottos);
```
totalLottos를 구하는 로직 자체가 비즈니스를 담고있어서, main이 아닌 도메인이 가지고 있는게 더 좋을 것 같습니다. :)

stopsilver13 replied:
메서드로 따로 분리하였습니다!

phs1116 reviewed (changes requested):
안녕하세요. 리뷰어 화투입니다.
피드백 잘 받아주셔서 감사합니다. 💯 
잘 구현해 주신것 같아요.
간단한 피드백 몇가지 추가했어요.
확인 부탁드리고 궁금하신 점 있으시면 언제든지 DM주세요!

Commit:
[REVIEW] Use interface to refer the LottoResult > lottoResult property

> 이펙티브자바64 객체는 인터페이스를 사용해 참조하라

Commit:
[REFACTOR] Fix bug that match any rank if bonus match and not SECOND

> countOfMatch와 bonusMatch를 and조건으로 걸어서
  SECOND rank 가 아닐 때 rank 매칭 안되는 버그 수정

Commit:
[REVIEW] Delete unnecessary method by using getOrDefault

Commit:
[REVIEW] Use interface to refer the LottoCreator

> 이펙티브자바64 객체는 인터페이스를 사용해 참조하라

Commit:
[REVIEW] Extract logic from main to other method

> total lottos로 합치는 로직 메서드로 분리

Commit:
[REFACTOR] Move responsibility of get random numbers

> LottoNumber 에서 AutoLottoCreator 로 책임이전

Commit:
[REFACTOR] Move responsibility of check is lotto number

> ManualLottoCreator 에서 LottoNumber 로 책임이전

Commit:
[REVIEW] Reduce dependency on LottoCreator for LottoFactory

> LottoFactory가 로또 생성시 LottoCreator list를 주입받도록 변경
> Creator를 갯수만큼 생성해주는 CreatorFactory 생성
> 구입한 로또 출력받는 뷰 구현

phs1116 reviewed (approved):
안녕하세요! 리뷰어 화투입니다.
주고받은 의견에 대해서 진짜 많이 고민한 흔적이 보이네요! 👍 
이번 단계 머지할게요 :)
구현하시느라 너무 고생 많으셨습니다.
